Speakeasy Two Factor Authentication in Nodejs (#2046)
section-engineering-education/engineering-education: articles/speakeasy-two-factor-authentication-in-nodejs/index.md

### Goal
In this tutorial, we will do a backend implementation of two-factor authentication using a token generated by the google authenticator app.

### Table of contents
- [Introduction](#introduction)
- [Setting up the Node application](#setting-up-the-nodejs-application)
- [Installing the dependencies](#installing-the-dependencies)
- [Setting up the application](#setting-up-the-application)
- [The node json databse](#the-node-json-databse)
- [Sending requests using postman](#sending-requests-using-postman)
- [Registering users](#registering-users)
- [Verifying the user](#verifying-the-user)
  - [Retriving user-id and temp secret from the database](#retrieving-id-and-temp-secret-from-the-database)
  - [Generating verification token](#generating-verification-token)
  - [Sending post request for verification response via postman](#sending-post-request-for-verification-response-via-postman)
- [Validate the token ](#validate-the-token )
- [Running the server](#running-the-server)
- [Conclusion](#conclusion)
- [Further Reading](#further-reading)
### Introduction

> The Speakeasy library provides two-factor authentication using a one-time pin(OTP). The library provides an additional layer of security to an application's standards authentication process. Using the OTP, Speakeasy provides extra data required for account access.

Reviewer: Add this section before `###Goal`.

Reviewer: Firstly, I would suggest you to tell how authentication happens? Then, give a little intro about two-factor authentication. Later, you could explain more about Speakeasy library.

Also, add a link to the repo or the npm package of this library

### Setting up the nodejs application
Let us begin by initializing our application using the `init` command.
```bash
npm init -y
```

> The command will create a package.json file for our app holding the metadata about the project.

Reviewer: package.json -> `package.json`.

For filenames, variables, constants, you can highlight them using ` `` `

```json
{
  "name": "node-two-factor-auth",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1",
    "start": "nodemon start"
  },
  "keywords": [],
  "author": "mia-roberts",
  "license": "ISC",
  "dependencies": {
    "express": "^4.17.1",
    "node-json-db": "^1.3.0",
    "speakeasy": "^2.0.0",
    "uuid": "^8.3.2"
  }
}
```

### Installing the dependencies
After the setup, our application needs some project-specific dependencies. The dependencies include:
+ `Express` as the backend server.
+ `uuid` to generate universal unique user-id.
+ `node-json-db` as our database to store the data as json.

> + `speakeasy` library for authetication

Reviewer: ```suggestion
+ `speakeasy` library for authentication
```

> + `nodemon` for use as our development dependency. With nodemon, we won't have to restart our application every time we make a change.

Reviewer: For bullet points, do not use `+`, instead use `-`

```js
//adding the speakeasy library
const speakeasy = require('speakeasy')

//adding the uuid dependacy
const uuid = require('uuid')

//adding node-json-db
const { JsonDB } = require('node-json-db')
const { Config } = require('node-json-db/dist/lib/JsonDBConfig')
```

### The node json databse
Our application will use the node json database to store user records in json format. Check out  [this link](https://www.npmjs.com/package/node-json-db?S_TACT=AGRAK616). To find more about `node-json-db`. From the documentation of `node-json-db`, to initialize the `node-json-db`, we will add the script below.
```js

// instance of the node-json-db
const db = new JsonDB(new Config("DataBase", true, false, '/'));
```
 * `new Config`  - Creates a node-json-db database config
 * `DataBase` - specifies the name of the JSON storage file
 * `true` Tells the database to save on each push.
 * `false` Instructs the database to save the database in a human-readable format

>  * `/` - the separator to use when accessing  our database values

Reviewer: Use `-`

### Sending requests using postman

> Since our application does not have a frontend, we will use postman for sending the requests to the applications' backend. Postman provides an interface for handling requests that would have otherwise been handled by the HTML. Check out [this guide](https://www.blazemeter.com/blog/how-use-postman-manage-and-execute-your-apis) on how to use postman. You can download postman using [this](https://www.postman.com/downloads/) link. 

Reviewer: ```suggestion
Since our application does not have a frontend, we will use postman for sending the requests to the applications' backend.

Postman provides an interface for handling requests that would have otherwise been handled by the HTML.

Check out [this guide](https://www.blazemeter.com/blog/how-use-postman-manage-and-execute-your-apis) on how to use postman. You can download postman using [this](https://www.postman.com/downloads/) link. 
```

> In postman, we will use three routes, register, verify and validate route. So we will create the URLs as below:

Reviewer: ```suggestion
In postman, we will use three routes, `/register`, `/verify`, and `/validate` route. So we will create the URLs as below:
```

- Register: http://localhost:5000/api/register
- Verify: http://localhost:5000/api/verify
- Validate: http://localhost:5000/api/validate

### Registering users

> In this application, we assume that a user is authetcated using his/her user-id. We therefore ignore other user details. We will resgister users and store their `user-id`  alongside the `secret-key` generated by speakeasy in the `Database.json` file. The registration process  begins by passing a post request to the `register` route in the `index.js` file. We then use `uuid` to generate a unique `user-id` and generate a temporary `secret-key` for the user ID. Next, we store the `user-id` and the `secret-key` in the `node-json-db`. The code for the process is as below:

Reviewer: ```suggestion
In this application, we assume that a user is authenticated using his/her user-id. We therefore ignore other user details. We will register users and store their `user-id`  alongside the `secret-key` generated by speakeasy in the `Database.json` file.

The registration process begins by passing a post request to the `register` route in the `index.js` file. We then use `uuid` to generate a unique `user-id` and generate a temporary `secret-key` for the user ID.

Next, we store the `user-id` and the `secret-key` in the `node-json-db`.

The code for the process is as below:
```

Reviewer: Always break down bigger paragraphs into smaller ones to enhance the readability.

One paragraph must speak only about one idea

    try {
        const path = `user/${userid }` 
        const temp_secret = speakeasy.generateSecret()
        db.push(path,{ userid , temp_secret }) 
        // we only need the base32 
        response.json({userid , secret: temp_secret.base32})
    }catch (error) {
        console.log(error)
        response.status(500).json({message: 'Error generating the secret'})
    }
})
```
The user object in the database will be as shown below:
```json
{
    "user":{
        "00e296df-cff6-44ee-94f7-763de86962c3":{
            "id":"00e296df-cff6-44ee-94f7-763de86962c3",
            "temp_secret":{
                "ascii":"eez>9svVgNa$DE9TXZQw#z0dkXI!GSQT",
                "hex":"65657a3e39737656674e612444453954585a5177237a30646b58492147535154",
                "base32":"MVSXUPRZON3FMZ2OMESEIRJZKRMFUULXEN5DAZDLLBESCR2TKFKA",
                "otpauth_url":"otpauth://totp/SecretKey?secret=MVSXUPRZON3FMZ2OMESEIRJZKRMFUULXEN5DAZDLLBESCR2TKFKA"
            }
        }
    }
}
```
### Verifying the user
Next, we need to verify our registered users using their id and temp-secret. We also need to make the secret permanent in the database. 
#### Retrieving id and temp secret from the database

> Since we will need the user-id and temp secret, we extract them from the database using the below code:

Reviewer: ```suggestion
Since we will need the `user-id` and `temp_secret`, we extract them from the database using the below code:
```

    const {token, userId} = req.body;   
    try {
      // Retrieve user from database
      const path = `/user/${userId}`;
      const user = db.getData(path);
    
      const { base32: secret } = user.secret;
      const tokenValidate = speakeasy.totp.verify({
        secret,
        encoding: 'base32',
        token, 
        window:1 // time window
      });
    
      if (tokenValidate) {
        
        res.json({ validated: true })
      } else {
        res.json({ validated: false})
      }
    }catch(error) {
      console.error(error)
      res.status(500).json({ message: "Error retrieving user!"})
    };
  })
```
To check if a user is validated, we will navigate to the validate route in postman, supply the user id and token, then submit the post request to the application to call the above block of code.

### Running the server
To test out the application, we add the block of code below in the index.js file. the run the command `npm start` in your terminal.
```js
const PORT = process.env.PORT || 5000
app.listen(PORT, () =>{
    console.log(`Server running on port ${PORT}`);
})
```
The images below show the results for each request sent via postman:

> 1. ![Register route](/engineering-education/speakeasy-two-factor-authentication-in-nodejs/register-route.png)
> 1. ![Verify route](/engineering-education/speakeasy-two-factor-authentication-in-nodejs/verify-route.png)
> 2. ![Validate route](/engineering-education/speakeasy-two-factor-authentication-in-nodejs/validate-route.png)

Reviewer: ```suggestion
![Register route](/engineering-education/speakeasy-two-factor-authentication-in-nodejs/register-route.png)
*Register route*

![Verify route](/engineering-education/speakeasy-two-factor-authentication-in-nodejs/verify-route.png)
*Verify route*

![Validate route](/engineering-education/speakeasy-two-factor-authentication-in-nodejs/validate-route.png)
*Validate route*
```

### Conclusion

> We set up a node application and coded a backend for two-factor authetication using speakeasy. We also learned how to use the authenticator extension in chrome for generation tokens. Lastly, we used postman to simulate sending requests from the front end of an application. Check the links in the section below for more about the topic. You can find the code for implementation [here.](https://replit.com/@miaroberts1/nodejs-2FA-using-speakeasy) You will, however, have to download the source code and install postman for testing.

Reviewer: ```suggestion
We set up a node application and coded a backend for two-factor authetication using speakeasy. We also learned how to use the authenticator extension in chrome for generation tokens. Lastly, we used postman to simulate sending requests from the front end of an application.

Check the links in the section below for more about the topic. You can find the code for implementation [here](https://replit.com/@miaroberts1/nodejs-2FA-using-speakeasy). You will, however, have to download the source code and install postman for testing.
```
